### PR TITLE
Homework modern cpp #2

### PR DESCRIPTION
**Pull request by DamiXioR**

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Good but I think that delegating constructors are missing

File: shapes/Circle.cpp [hunk 1]
```diff
@@ -6,6 +6,11 @@ Circle::Circle(double r)
     : r_(r)
```
Comment (ziobron) [inline shapes/Circle.cpp]: no constructor delegation. It takes place when one constructor calls another of the same class on initializer list.

File: shapes/Rectangle.cpp [hunk 2]
```diff
@@ -1,10 +1,23 @@
...
 
+Rectangle::Rectangle(Color c)
+    : Shape(c)
+{
+}
+
 Rectangle::Rectangle(double x, double y)
     : x_(x),
       y_(y)
-{}
```
Comment (ziobron) [inline shapes/Rectangle.cpp]: constructor delegation missing

File: shapes/Square.cpp [hunk 3]
```diff
@@ -1,11 +1,15 @@
 #include "Square.hpp"
 #include <iostream>
 
-Square::Square(Color c) : Rectangle(c) {}
-
 Square::Square(double x)
     : Rectangle(x, x)
```
Comment (ziobron) [inline shapes/Square.cpp]: constructor delegation is missing

File: shapes/main.cpp [hunk 4]
```diff
@@ -1,80 +1,90 @@
...
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
+//TASK #11
+template <class DerivedType>
+using SFINAE = std::enable_if<std::is_base_of<Circle, std::remove_reference_t<DerivedType>>::value>;
+
+template <class DerivedType, class = SFINAE<DerivedType>>
+auto insertSubclassOfShapes(Collection& collection, std::shared_ptr<DerivedType> shape)
 {
-    if(first == nullptr || second == nullptr)
+    auto insertPosition = collection.emplace_back(shape);
```
Comment (ziobron) [inline shapes/main.cpp]: not position, a reference to the inserted element is returned.

File: shapes/main.cpp [hunk 5]
```diff
@@ -1,80 +1,90 @@
...
-#include "Circle.hpp"
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
+//TASK #11
+template <class DerivedType>
+using SFINAE = std::enable_if<std::is_base_of<Circle, std::remove_reference_t<DerivedType>>::value>;
```
Comment (ziobron) [inline shapes/main.cpp]: Hmm... Circle? We want to enable this overload for all classes that inherit from Shape, not Circle.
```suggestion
using SFINAE = std::enable_if<std::is_base_of<Shape, std::remove_reference_t<DerivedType>>::value>;
```

File: shapes/main.cpp [hunk 6]
```diff
@@ -1,80 +1,90 @@
...
+
+template <class DerivedType, class = SFINAE<DerivedType>>
+auto insertSubclassOfShapes(Collection& collection, std::shared_ptr<DerivedType> shape)
```
Comment (ziobron) [inline shapes/main.cpp]: shared_ptr can be taken by const &

File: shapes/main.cpp [hunk 7]
```diff
@@ -103,8 +136,44 @@ int main()
...
+    auto fastSquare = make_shape<Square>(20.7);
+    auto fastCircle = make_shape<Circle>(4, Color::Black);
+    auto fastRectangle = make_shape<Rectangle>(44.4, 20.7, Color::Green);
+    fastSquare->getColor();
+    fastCircle->getColor();
+    fastRectangle->getColor();
+
+    //TASK #11
+    std::cout << "TASK #11\n";
+    auto insertedIn = insertSubclassOfShapes<Rectangle>(shapes, std::make_shared<Rectangle>(50, 40, Color::Black));
```
Comment (ziobron) [inline shapes/main.cpp:176]: does it compile?

**Review by ziobron — CHANGES_REQUESTED** [review]
There is a CI error: https://travis-ci.com/github/coders-school/modern-cpp/jobs/423283338

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have nothing to add that was not mentioned by previous reviewers.

File: shapes/main.cpp (hunk 7 above)
Comment (DamiXioR) [reply shapes/main.cpp:176]: i think it works now.

File: shapes/Circle.cpp [hunk 8]
```diff
@@ -2,10 +2,6 @@
 #include <math.h>
 #include <iostream>
 
-Circle::Circle(double r)
-    : r_(r)
-{}
```
Comment (ziobron) [inline shapes/Circle.cpp]: You have removed this constructor, but you should use a delegation. I know that default argument version has the same behavior, but the idea here was to show constructor delegation feature.
You should leave this constructor and call below constructor on initializer list.
The same for other shapes.

**Review by ziobron — CHANGES_REQUESTED** [review]

**Comment by DamiXioR** [comment]
Its everything allright :)?

**Comment by ziobron** [comment]
I'm waiting for you to add a constructor delegation. Take a look at other's people solutions. Constructor delegation takes place when you call another constructor of the same class on initialization list.

**Comment by DamiXioR** [comment]
> I'm waiting for you to add a constructor delegation. Take a look at other's people solutions. Constructor delegation takes place when you call another constructor of the same class on initialization list.

I think it's done.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**